dbft: basic payloads implementation

AnnaShaleva opened the pull request:
Close #13. Close #31. Implement a part of #29.

In this PR:
* Introduce custom dBFT payloads implementation;
* Integrate DBFT service with dBFT message pool;
* Integrate DBFT service with transactions mempool and chain events;
* Introduce basic dBFT verification rules;
* Make multiple consensus nodes accept blocks, accept transactions.

A set of related issues are created for further work. Several critical issues exist in the current implementation that under certain circumstances may stop consensus or kill the node, they will be resolved in separate PRs.

AnnaShaleva committed:
dbft: basic payloads implementation

A final major part of #13. Introduce custom dBFT payloads implementation,
integrate DBFT service with dBFT message pool, introduce basic payloads
verification rules.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

roman-khimov commented on consensus/dbft/prepare_request.go:
```diff
@@ -0,0 +1,99 @@
+	// TODO: may be removed, as we have sealingProposal's timestamp OR handle with care, because it's dBFT
+	// who sets this to custom value from context.
+	timestamp uint64
+	// Do we really need nonce?
```
I see you have a full header below, so likely it's not needed.

roman-khimov commented on consensus/dbft/prepare_request.go:
```diff
@@ -0,0 +1,99 @@
+	// version uint32
+	// timestamp is a nanoseconds-precision timestamp, exactly like dBFT operates with.
+	// TODO: may be removed, as we have sealingProposal's timestamp OR handle with care, because it's dBFT
+	// who sets this to custom value from context.
```
Maybe this one as well, if you're reusing E-header.

roman-khimov commented on consensus/dbft/prepare_request.go:
```diff
@@ -0,0 +1,99 @@
+		w.Err = fmt.Errorf("failed to encode PrepareRequest to RLP: %w", err)
+		return
+	}
+	w.WriteVarBytes(b)
```
I don't like mixing N3-style encoding and RLP, but we can create a separate issue for that, N3 is easier for initial port.

AnnaShaleva replied:
Will create an issue for it.

AnnaShaleva replied:
Ref. https://github.com/bane-labs/go-ethereum/issues/62.

roman-khimov commented on consensus/dbft/signer.go:
```diff
@@ -22,5 +22,5 @@ type Signer struct {
 // guaranteed by Block.Sign method.
 func (s *Signer) Sign(msg []byte) ([]byte, error) {
-	return s.SignFn(accounts.Account{Address: s.Signer}, accounts.MimetypeClique, msg)
+	return s.SignFn(accounts.Account{Address: s.Signer}, accounts.MimetypeClique, msg) // TODO: replace this accounts.MimetypeClique?
```
It was mentioned in #12, either we need a separate issue now or deal with it here.

AnnaShaleva replied:
A separate issue, will create.

AnnaShaleva replied:
Ref. https://github.com/bane-labs/go-ethereum/issues/61.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -705,30 +797,41 @@ func (c *DBFT) verifySeal(snap *Snapshot, header *types.Header, parents []*types
+	}
+	/*
+		// TODO: get rid of difficulty?
+		// Ensure that the difficulty corresponds to the turn-ness of the signer
```
IIRC (but not sure) in recent versions it's related to fee calculations, so maybe it's still needed. Do you have any problem with it?

AnnaShaleva replied:
We have a problem, but it's not critical, will create a separate issue for it.

AnnaShaleva replied:
Ref. #58.

AnnaShaleva committed:
dbft: refactor processNewBlock callback

Create a notion of block queuer, move dBFT block awaiting task from
Seal to processNewBlock and block queuer, do not explicitly hold
Seal for the whole dBFT consensus round.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: ensure last proposal matches the one Seal is called for

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
protocols: optimize dBFT payload verification

Just a bit, do not perform intermediate steps.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: allow to subscribe for chain and mempool

Subscribe for transactions to route them to dBFT and subscribe for
block headers to notify consensus about new blocks in chain. A part
of #29.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: single privnet workaround

Just make the single-node consensus not to stop blocks acceptance after
the first accepted block. It's a kludge and likely will be replaced in
future.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: adjust local prepareRequest implementation for proper rlp encoding

Unexported fields can't be properly RLP-encoded.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: implement dBFT block signature verification

Once new Commit message is accepted by node, it's being verified against
the currently stored proposal using this function.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: dettach payloadFromMessage from DBFT service

This method doesn't require DBFT instance to work properly.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
protocols: do not recalculate hash while iterating peers with dBFT message

A tiny optimisation, hash may be compted once.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: improve logging level

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: do not leak payload abstraction in newPayload dBFT callback

If it's PrepareRequest, then the whole PrepareRequest is properly initialized
via WithNewPrepareRequest constructor.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: do not decline sealing in case of rescently sealed block

DBFT is not a miner, there's no "rescent signers", dBFT must be
started every block and it must have proposal under it. This commit
prevents the situation when primary of the previous round can't start
new sealing in the next round:
```
2023-12-14T15:59:57.028+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507	received Commit	{"validator": 3}
2023-12-14T15:59:57.028+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:66	approving block	{"height": 1, "hash": "ed0240024dc5c554cab865b8349a780d569e50a9a8aeb19893eab11353ebae4a", "tx_count": 0, "merkle": "7a77851842736c8672a67c64093fcdae5e1991f8cee5806a2d80a4b70a9749a9", "prev": "c17832c1d0b3cc7b231c1effd124c7d956d7de3cb31c3e5eceda0a55c8f7f453"}
2023-12-14T15:59:57.029+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 2, "height": 1, "view": 0, "my_height": 1, "my_view": 0}
INFO [12-14|15:59:57.031] Successfully sealed new block            number=1 sealhash=ed0240..ebae4a hash=afc8b6..288bd2 elapsed=15.011s
INFO [12-14|15:59:57.032] Commit new sealing work                  number=2 sealhash=379d37..684c8a txs=0 gas=0 fees=0 elapsed="301.639µs"
WARN [12-14|15:59:57.032] Block sealing failed                     err="signed recently, must wait for others"
WARN [12-14|15:59:57.033] Snap syncing, discarded propagated block number=1 hash=c9127a..1c567f
WARN [12-14|15:59:57.035] Snap syncing, discarded propagated block number=1 hash=c9127a..1c567f
2023-12-14T16:00:37.032+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "RecoveryRequest", "from": 3, "height": 2, "view": 0, "my_height": 1, "my_view": 0}
2023-12-14T16:00:37.032+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:229	caching message from future	{"height": 2, "view": 0, "cache": null}
```

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: drop snapshot's Recents mentions

No spam is possible with dBFT, thus, remove all mentions of rescent signers.
Prevent the following error after block acceptance:
```
2023-12-14T16:58:12.519+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:57	not enough to commit	{"count": 2}
2023-12-14T16:58:12.519+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 0, "height": 2, "view": 1, "my_height": 2, "my_view": 1}
2023-12-14T16:58:12.519+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507	received Commit	{"validator": 0}
2023-12-14T16:58:12.519+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:66	approving block	{"height": 2, "hash": "6c8ed4b751c503e66eb526d1b0c21fed352ae62cd5d7c81e7885a974fb9a5493", "tx_count": 0, "merkle": "7a77851842736c8672a67c64093fcdae5e1991f8cee5806a2d80a4b70a9749a9", "prev": "517278baa2396dcfa198e763df96ddfbd548fb26ad8854f378d4234d06b973ad"}
INFO [12-14|16:58:12.531] Successfully sealed new block            number=2 sealhash=6c8ed4..9a5493 hash=9224b6..6d21a1 elapsed=2m15.022s
ERROR[12-14|16:58:12.531] Failed to prepare header for sealing     err="failed to create snapshot for voting calculations: recently signed"
INFO [12-14|16:58:21.594] Looking for peers                        peercount=3 tried=0 static=0
INFO [12-14|16:58:31.614] Looking for peers                        peercount=3 tried=0 static=0
```

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: use buffered channel for chain blocks subscription

Receive from blocks subscription channel is happened at the same routine as
handling dBFT messages, thus, dBFT event loop may be blocked on subsequent
commit if this channel is not buffered.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: allow to insert dBFT block into chain on missing sealing task

It's possible that proposed dBFT block differs from the sealing task provided
by miner. The block itself is still valid, thus, insert it directly to the chain.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: improve error logging

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
worker: change fork-tolerant block number threshold

No forks, single-block finality. If we have some block received from chain
then miner have to stop sealing tasks for that height.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: allow forks (blocks with same index and different signs)

Block's Extra part (last signatures-related bytes) may differ from node
to node. We need to allow such forks on the consensus-level, thus, consider
subsequent PrepareRequest valid if it contains proper parent's SealHash
and save new parent with updated extra.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

roman-khimov reviewed (commented):
I'm somewhat worried about locks and queues here, but it looks legit otherwise.

AnnaShaleva committed:
dbft: reorder postBlock callback and sending block to chain

Otherwise race is possible between postBlock and new sealing task which
in bad case leads to the task cancellation.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: refactor (*blockQueue).PutBlock

Firstly, try to search for a suitable SealTask and if it's found, then
we're lucky and no additional actions are needed. If there's no pending
matching sealing task or worker has cancelled the task, then check the
block in chain and insertt it if it's missing from chain.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: fix typo

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: remove stale tasks from blockqueue

We don't expect a lot of pending tasks to be present in the blockQueue,
it's likely a single or a couple of them at once. But we need to clear
unused tasks eventually.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: add an ability to retrieve missing transaction from mempool

Subscribe for mempooled transactions and retrieve missing transactions
from pool if needed.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: adjust newBlockFomContext callback

We have ParentHash in dBFT proposal (prepare request), thus, it's already
properly set via `sealingProposal`.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: rework NewBlockFromContext and ProcessBlock callbacks

NewBlockFromContext must operate with a pure context and PrepareRequest,
in particular. It must not include transactions, because some of the
transactions may be missing by the moment of call to NewBlockFromContext.
Close #31.

dBFT-consistent Block structure should not include types.Block itself,
it's sufficient to have types.Header and a set of transactions.

ProcessBlock callback should create new block from the proposed header
by adding signature to the header Extra and by adding transactions to the
resulting types.Block.

Avoid creating a lot of block copies, avoid usage of the whole Block where
it's sufficient to have only Header.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
> I'm somewhat worried about locks and queues here, but it looks legit otherwise.

I also worried about locks, and I've already caught a couple of connected bugs, but it's just the way how dBFT performs synchronisation with miner. See the updated version, I've tried to remove locks from dBFT callbacks as much as possible, and it looks a bit better now.

AnnaShaleva commented:
OK, let me share the progress update wrt the current PR:

First of all, four-nodes privnet is running, here's the example of node log:
```
2023-12-15T18:42:06.262+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:57	not enough to commit	{"count": 2}
2023-12-15T18:42:06.262+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 1, "height": 35, "view": 0, "my_height": 35, "my_view": 0}
2023-12-15T18:42:06.262+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507	received Commit	{"validator": 1}
WARN [12-15|18:42:06.262] Snap syncing, discarded propagated block number=35 hash=7d9d34..7451db
2023-12-15T18:42:06.262+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:66	approving block	{"height": 35, "hash": "a765aa614b219b0ed0507bff18d1cab4f09576cf1540a3d2dcccc029e2c1ea6f", "tx_count": 0, "merkle": "7a77851842736c8672a67c64093fcdae5e1991f8cee5806a2d80a4b70a9749a9", "prev": "9b93ff88cbfbbe5e248858506253d6181c0ebc5aee92710d3dc367e36a734d46"}
INFO [12-15|18:42:06.265] Chain reorg detected                     number=33 hash=f6958e..dcc81c drop=1 dropfrom=b11223..820fde add=2 addfrom=0305f9..1c1e79
INFO [12-15|18:42:06.267] Imported new chain segment               number=35 hash=0305f9..1c1e79 blocks=1 txs=0 mgas=0.000 elapsed=4.961ms     mgasps=0.000 triedirty=0.00B
2023-12-15T18:42:06.268+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 35, "view": 0, "my_height": 35, "my_view": 0}
2023-12-15T18:42:06.268+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 1, "height": 35, "view": 0, "my_height": 35, "my_view": 0}
2023-12-15T18:42:06.268+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 35, "view": 0, "my_height": 35, "my_view": 0}
INFO [12-15|18:42:06.269] Commit new sealing work                  number=36 sealhash=ac2577..808f03 txs=0 gas=0 fees=0 elapsed="781.355µs"
2023-12-15T18:42:06.269+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:107	initializing dbft	{"height": 36, "view": 0, "index": 0, "role": "Primary"}
2023-12-15T18:42:06.269+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 36, "v": 0, "delay": "4.993425736s"}
2023-12-15T18:42:11.263+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:185	timeout	{"height": 36, "view": 0}
2023-12-15T18:42:11.263+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:9	broadcasting message	{"type": "PrepareRequest", "height": 36, "view": 0}
2023-12-15T18:42:11.264+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:40	sending PrepareRequest	{"height": 36, "view": 0}
2023-12-15T18:42:11.264+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 36, "v": 0, "delay": "5s"}
2023-12-15T18:42:11.264+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:29	check preparations	{"hasReq": true, "count": 1, "M": 3}
2023-12-15T18:42:11.321+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareResponse", "from": 1, "height": 36, "view": 0, "my_height": 36, "my_view": 0}
2023-12-15T18:42:11.322+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:433	received PrepareResponse	{"validator": 1}
2023-12-15T18:42:11.322+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:29	check preparations	{"hasReq": true, "count": 2, "M": 3}
2023-12-15T18:42:11.322+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareResponse", "from": 3, "height": 36, "view": 0, "my_height": 36, "my_view": 0}
2023-12-15T18:42:11.322+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:433	received PrepareResponse	{"validator": 3}
2023-12-15T18:42:11.322+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:29	check preparations	{"hasReq": true, "count": 3, "M": 3}
2023-12-15T18:42:11.322+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:134	sending Commit	{"height": 36, "view": 0}
2023-12-15T18:42:11.322+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:9	broadcasting message	{"type": "Commit", "height": 36, "view": 0}
2023-12-15T18:42:11.323+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 36, "v": 0, "delay": "5s"}
2023-12-15T18:42:11.323+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:57	not enough to commit	{"count": 1}
2023-12-15T18:42:11.325+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareResponse", "from": 2, "height": 36, "view": 0, "my_height": 36, "my_view": 0}
2023-12-15T18:42:11.325+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:433	received PrepareResponse	{"validator": 2}
2023-12-15T18:42:11.327+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 2, "height": 36, "view": 0, "my_height": 36, "my_view": 0}
2023-12-15T18:42:11.327+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507	received Commit	{"validator": 2}
2023-12-15T18:42:11.328+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:57	not enough to commit	{"count": 2}
2023-12-15T18:42:11.328+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 1, "height": 36, "view": 0, "my_height": 36, "my_view": 0}
2023-12-15T18:42:11.328+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507	received Commit	{"validator": 1}
2023-12-15T18:42:11.329+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:66	approving block	{"height": 36, "hash": "ac2577a1ec758adbcb9c7ace535a5f8315d00514979e795e6c1b84cbda808f03", "tx_count": 0, "merkle": "7a77851842736c8672a67c64093fcdae5e1991f8cee5806a2d80a4b70a9749a9", "prev": "0305f9e1b9226d24606878ea9860870576337ed05eb63e095b35dc07a71c1e79"}
2023-12-15T18:42:11.334+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 36, "view": 0, "my_height": 36, "my_view": 0}
INFO [12-15|18:42:11.334] Successfully sealed new block            number=36 sealhash=ac2577..808f03 hash=4c6d03..422691 elapsed=5.065s
2023-12-15T18:42:11.334+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 36, "view": 0, "my_height": 36, "my_view": 0}
INFO [12-15|18:42:11.335] Commit new sealing work                  number=37 sealhash=809839..925d8d txs=0 gas=0 fees=0 elapsed="951.533µs"
2023-12-15T18:42:11.335+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:107	initializing dbft	{"height": 37, "view": 0, "index": 0, "role": "Backup"}
2023-12-15T18:42:11.335+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 37, "v": 0, "delay": "9.993129488s"}
2023-12-15T18:42:11.338+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 36, "view": 0, "my_height": 37, "my_view": 0}
2023-12-15T18:42:11.338+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:223	ignoring old height	{"height": 36}
INFO [12-15|18:42:14.962] Looking for peers                        peercount=3 tried=0 static=0
2023-12-15T18:42:16.332+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareRequest", "from": 1, "height": 37, "view": 0, "my_height": 37, "my_view": 0}
2023-12-15T18:42:16.332+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:325	received empty PrepareRequest
2023-12-15T18:42:16.332+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:333	received PrepareRequest	{"validator": 1, "tx": 0}
2023-12-15T18:42:16.332+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:106	sending PrepareResponse	{"height": 37, "view": 0}
2023-12-15T18:42:16.332+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:9	broadcasting message	{"type": "PrepareResponse", "height": 37, "view": 0}
2023-12-15T18:42:16.332+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:29	check preparations	{"hasReq": true, "count": 2, "M": 3}
2023-12-15T18:42:16.388+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareResponse", "from": 2, "height": 37, "view": 0, "my_height": 37, "my_view": 0}
2023-12-15T18:42:16.388+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:433	received PrepareResponse	{"validator": 2}
2023-12-15T18:42:16.388+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:29	check preparations	{"hasReq": true, "count": 3, "M": 3}
2023-12-15T18:42:16.389+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:134	sending Commit	{"height": 37, "view": 0}
2023-12-15T18:42:16.389+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:9	broadcasting message	{"type": "Commit", "height": 37, "view": 0}
2023-12-15T18:42:16.391+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 37, "v": 0, "delay": "5s"}
2023-12-15T18:42:16.391+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:57	not enough to commit	{"count": 1}
2023-12-15T18:42:16.391+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareResponse", "from": 3, "height": 37, "view": 0, "my_height": 37, "my_view": 0}
2023-12-15T18:42:16.391+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:433	received PrepareResponse	{"validator": 3}
2023-12-15T18:42:16.391+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareResponse", "from": 3, "height": 37, "view": 0, "my_height": 37, "my_view": 0}
2023-12-15T18:42:16.391+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:425	ignoring PrepareResponse because of view changing
2023-12-15T18:42:16.391+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 2, "height": 37, "view": 0, "my_height": 37, "my_view": 0}
2023-12-15T18:42:16.391+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507	received Commit	{"validator": 2}
2023-12-15T18:42:16.391+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:57	not enough to commit	{"count": 2}
2023-12-15T18:42:16.391+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareResponse", "from": 3, "height": 37, "view": 0, "my_height": 37, "my_view": 0}
2023-12-15T18:42:16.391+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:425	ignoring PrepareResponse because of view changing
2023-12-15T18:42:16.391+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 37, "view": 0, "my_height": 37, "my_view": 0}
2023-12-15T18:42:16.391+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507	received Commit	{"validator": 3}
2023-12-15T18:42:16.392+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:66	approving block	{"height": 37, "hash": "8098391340bcff15fe72dbd165e29b8478ad539802d4eb96fad3d52e4e925d8d", "tx_count": 0, "merkle": "7a77851842736c8672a67c64093fcdae5e1991f8cee5806a2d80a4b70a9749a9", "prev": "4c6d0342db71b52e5eec01a65fa23e84656b26ad6135de2813a4ef59e9422691"}
2023-12-15T18:42:16.392+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 1, "height": 37, "view": 0, "my_height": 37, "my_view": 0}
WARN [12-15|18:42:16.395] Snap syncing, discarded propagated block number=37 hash=7e83de..aad9d8
INFO [12-15|18:42:16.395] Successfully sealed new block            number=37 sealhash=809839..925d8d hash=7aa5c0..8df0f8 elapsed=5.060s
INFO [12-15|18:42:16.395] Commit new sealing work                  number=38 sealhash=70f28c..71b5e9 txs=0 gas=0 fees=0 elapsed="168.017µs"
2023-12-15T18:42:16.395+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:107	initializing dbft	{"height": 38, "view": 0, "index": 0, "role": "Backup"}
2023-12-15T18:42:16.395+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 38, "v": 0, "delay": "9.996149933s"}
2023-12-15T18:42:21.395+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareRequest", "from": 2, "height": 38, "view": 0, "my_height": 38, "my_view": 0}
2023-12-15T18:42:21.395+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:325	received empty PrepareRequest
2023-12-15T18:42:21.395+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:333	received PrepareRequest	{"validator": 2, "tx": 0}
2023-12-15T18:42:21.395+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:106	sending PrepareResponse	{"height": 38, "view": 0}
2023-12-15T18:42:21.395+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:9	broadcasting message	{"type": "PrepareResponse", "height": 38, "view": 0}
2023-12-15T18:42:21.396+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:29	check preparations	{"hasReq": true, "count": 2, "M": 3}
2023-12-15T18:42:21.399+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareResponse", "from": 1, "height": 38, "view": 0, "my_height": 38, "my_view": 0}
2023-12-15T18:42:21.399+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:433	received PrepareResponse	{"validator": 1}
2023-12-15T18:42:21.399+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:29	check preparations	{"hasReq": true, "count": 3, "M": 3}
2023-12-15T18:42:21.399+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:134	sending Commit	{"height": 38, "view": 0}
2023-12-15T18:42:21.399+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:9	broadcasting message	{"type": "Commit", "height": 38, "view": 0}
2023-12-15T18:42:21.399+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 38, "v": 0, "delay": "5s"}
2023-12-15T18:42:21.399+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:57	not enough to commit	{"count": 1}
2023-12-15T18:42:21.399+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 1, "height": 38, "view": 0, "my_height": 38, "my_view": 0}
2023-12-15T18:42:21.399+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507	received Commit	{"validator": 1}
2023-12-15T18:42:21.400+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:57	not enough to commit	{"count": 2}
2023-12-15T18:42:21.400+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 2, "height": 38, "view": 0, "my_height": 38, "my_view": 0}
2023-12-15T18:42:21.400+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507	received Commit	{"validator": 2}
2023-12-15T18:42:21.400+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:66	approving block	{"height": 38, "hash": "70f28ce903c0859f9670ae6ac99228cb566246b176361c23a24b36c23b71b5e9", "tx_count": 0, "merkle": "7a77851842736c8672a67c64093fcdae5e1991f8cee5806a2d80a4b70a9749a9", "prev": "7aa5c0eca68ded87354791a8cdaf09679277fa38916d738dff9fe45d148df0f8"}
2023-12-15T18:42:21.400+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 2, "height": 38, "view": 0, "my_height": 38, "my_view": 0}
2023-12-15T18:42:21.403+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareResponse", "from": 3, "height": 38, "view": 0, "my_height": 38, "my_view": 0}
INFO [12-15|18:42:21.404] Successfully sealed new block            number=38 sealhash=70f28c..71b5e9 hash=76a5a0..26f5ce elapsed=5.008s
INFO [12-15|18:42:21.404] Commit new sealing work                  number=39 sealhash=926bfd..c15195 txs=0 gas=0 fees=0 elapsed="96.367µs"
2023-12-15T18:42:21.404+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:107	initializing dbft	{"height": 39, "view": 0, "index": 0, "role": "Backup"}
2023-12-15T18:42:21.404+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 39, "v": 0, "delay": "9.996165128s"}
2023-12-15T18:42:21.404+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 38, "view": 0, "my_height": 39, "my_view": 0}
2023-12-15T18:42:21.404+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:223	ignoring old height	{"height": 38}
INFO [12-15|18:42:24.978] Looking for peers                        peercount=3 tried=0 static=0
2023-12-15T18:42:26.407+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareRequest", "from": 3, "height": 39, "view": 0, "my_height": 39, "my_view": 0}
INFO [12-15|18:42:26.465] Chain reorg detected                     number=37 hash=7aa5c0..8df0f8 drop=1 dropfrom=76a5a0..26f5ce add=1 addfrom=ca0832..438893
INFO [12-15|18:42:26.467] Imported new chain segment               number=38 hash=ca0832..438893 blocks=1 txs=0 mgas=0.000 elapsed=59.974ms    mgasps=0.000 triedirty=0.00B
2023-12-15T18:42:26.468+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:325	received empty PrepareRequest
2023-12-15T18:42:26.468+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:333	received PrepareRequest	{"validator": 3, "tx": 0}
2023-12-15T18:42:26.468+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:106	sending PrepareResponse	{"height": 39, "view": 0}
2023-12-15T18:42:26.468+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:9	broadcasting message	{"type": "PrepareResponse", "height": 39, "view": 0}
```

Secondly, some transactions may be pushed to the chain and accepted in subsequent blocks, here's the example of logs for blocks with 1 and 9 transactions:
```
2023-12-15T18:42:36.564+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:57	not enough to commit	{"count": 2}
2023-12-15T18:42:36.565+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 1, "height": 41, "view": 0, "my_height": 41, "my_view": 0}
2023-12-15T18:42:36.565+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507	received Commit	{"validator": 1}
2023-12-15T18:42:36.565+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:66	approving block	{"height": 41, "hash": "ee954d854113e7609abbd839291b1ce9c30917705f7eae1097aaa71a7993a033", "tx_count": 1, "merkle": "395f4e6601c53b95387419bd8145bbb5a92f5b004a9aecbdf16789163b867733", "prev": "676d7cd4048595ddd012285dacca2a6ccebe2c68d173d9ba81d42cdc6de545ac"}
INFO [12-15|18:42:36.569] Chain reorg detected                     number=39 hash=8ceb12..ad92a5 drop=1 dropfrom=fa8a45..0074d6 add=2 addfrom=5e143f..6fb42c
INFO [12-15|18:42:36.571] Imported new chain segment               number=41 hash=5e143f..6fb42c blocks=1 txs=1  mgas=0.021 elapsed=5.826ms     mgasps=3.604 snapdiffs=130.00B triedirty=1.50KiB
2023-12-15T18:42:36.572+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 41, "view": 0, "my_height": 41, "my_view": 0}
2023-12-15T18:42:36.572+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 41, "view": 0, "my_height": 41, "my_view": 0}
2023-12-15T18:42:36.572+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 41, "view": 0, "my_height": 41, "my_view": 0}
INFO [12-15|18:42:36.573] Commit new sealing work                  number=42 sealhash=38b1e7..fea583 txs=9  gas=189,000 fees=0.000189 elapsed=1.476ms
2023-12-15T18:42:36.574+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:107	initializing dbft	{"height": 42, "view": 0, "index": 0, "role": "Backup"}
2023-12-15T18:42:36.574+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 42, "v": 0, "delay": "9.991568693s"}
2023-12-15T18:42:41.568+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareRequest", "from": 2, "height": 42, "view": 0, "my_height": 42, "my_view": 0}
INFO [12-15|18:42:41.625] Chain reorg detected                     number=40 hash=676d7c..e545ac drop=1 dropfrom=5e143f..6fb42c add=1 addfrom=ce5577..a4ba05
INFO [12-15|18:42:41.630] Imported new chain segment               number=41 hash=ce5577..a4ba05 blocks=1 txs=1  mgas=0.021 elapsed=61.635ms    mgasps=0.341 snapdiffs=130.00B triedirty=1.50KiB
2023-12-15T18:42:41.630+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:333	received PrepareRequest	{"validator": 2, "tx": 9}
2023-12-15T18:42:41.630+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:106	sending PrepareResponse	{"height": 42, "view": 0}
2023-12-15T18:42:41.630+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:9	broadcasting message	{"type": "PrepareResponse", "height": 42, "view": 0}
2023-12-15T18:42:41.631+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:29	check preparations	{"hasReq": true, "count": 2, "M": 3}
2023-12-15T18:42:41.631+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareResponse", "from": 1, "height": 42, "view": 0, "my_height": 42, "my_view": 0}
2023-12-15T18:42:41.632+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:433	received PrepareResponse	{"validator": 1}
2023-12-15T18:42:41.632+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:29	check preparations	{"hasReq": true, "count": 3, "M": 3}
2023-12-15T18:42:41.632+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:134	sending Commit	{"height": 42, "view": 0}
2023-12-15T18:42:41.632+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:9	broadcasting message	{"type": "Commit", "height": 42, "view": 0}
2023-12-15T18:42:41.633+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 42, "v": 0, "delay": "5s"}
2023-12-15T18:42:41.633+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:57	not enough to commit	{"count": 1}
INFO [12-15|18:42:41.634] Commit new sealing work                  number=42 sealhash=2a8cf0..ba63c0 txs=9  gas=189,000 fees=0.000189 elapsed=3.822ms
WARN [12-15|18:42:41.635] Block sealing failed                     err="previous sealing is not yet finished"
2023-12-15T18:42:41.635+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareResponse", "from": 3, "height": 42, "view": 0, "my_height": 42, "my_view": 0}
2023-12-15T18:42:41.636+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:433	received PrepareResponse	{"validator": 3}
2023-12-15T18:42:41.636+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 42, "view": 0, "my_height": 42, "my_view": 0}
2023-12-15T18:42:41.636+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507	received Commit	{"validator": 3}
2023-12-15T18:42:41.636+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:57	not enough to commit	{"count": 2}
2023-12-15T18:42:41.639+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "PrepareResponse", "from": 3, "height": 42, "view": 0, "my_height": 42, "my_view": 0}
2023-12-15T18:42:41.639+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:425	ignoring PrepareResponse because of view changing
2023-12-15T18:42:41.639+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 42, "view": 0, "my_height": 42, "my_view": 0}
2023-12-15T18:42:41.639+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 1, "height": 42, "view": 0, "my_height": 42, "my_view": 0}
2023-12-15T18:42:41.639+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:507	received Commit	{"validator": 1}
2023-12-15T18:42:41.640+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:66	approving block	{"height": 42, "hash": "3ff878930030c5d9b42c00ca5f0e2a53823956765f1a91310e1ede2200da8a23", "tx_count": 9, "merkle": "5fb5839898e3b11892ec5b23516a759af9ad4878a30bd489b050c0b61b069d5e", "prev": "ce557797b33f958072a2c801fababc0ceb3888e05628f32b27c3b2d455a4ba05"}
INFO [12-15|18:42:41.650] Imported new chain segment               number=42 hash=5ea31f..4d40c2 blocks=1 txs=9  mgas=0.189 elapsed=9.985ms     mgasps=18.927 snapdiffs=261.00B triedirty=2.75KiB
2023-12-15T18:42:41.650+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:214	received message	{"type": "Commit", "from": 2, "height": 42, "view": 0, "my_height": 42, "my_view": 0}
INFO [12-15|18:42:41.651] Commit new sealing work                  number=43 sealhash=f69e50..4c6dce txs=0  gas=0       fees=0        elapsed="496.656µs"
2023-12-15T18:42:41.651+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:107	initializing dbft	{"height": 43, "view": 0, "index": 0, "role": "Backup"}
2023-12-15T18:42:41.651+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 43, "v": 0, "delay": "9.988114072s"}
WARN [12-15|18:42:41.653] Snap syncing, discarded propagated block number=42 hash=514249..f3e639
```

Thirdly, as we've discussed there's a delayed block hash finality, the delay is one block. Every PrepareRequest contains parent block sealhash which allows to properly run a subsequent consensus round and agree on the final parent block hash. Chain itself is OK with "floating" new block hash, see the "Chain reorg detected" in the logs above.

This PR may be taken to run and test four-node consensus (and even 7 node consensus, but we don't yet have Makefile target for it), but **this PR is still under development**. Please, consider PR finalisation and merge before submitting bugs. There is a set of known problems and TODOs in this PR, and we're fixing them currently to merge this PR ASAP. At the same time all basic consensus-related logic is implemented and multinode consensus is running.

AnnaShaleva committed:
dbft: require sealing task match cached last block information

The only place where cached last block information may be changed in dBFT
process is PrepareRequest validation callback. Otherwise it's not safe to
update these values.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: do not lock cached last block info

The only place where cached last block info values may be updated is
postBlock callback (and DBFT instance initialisation, but dBFT is not
running by this moment yet). postBlock callback may be called only as
a part of eventLoop routine, thus, no concurrent access may happen to
last block info.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: increase blockEvents channel capacity

Chain reorg (new block from P2P) may happen at the same time as new dBFT block
should be stored (in OnReceive handler) which prevents eventLoop from running
the next iteration. So 1 is for dBFT-generated block and 1 is for reorg-time
block handling.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: include proposed parent witness verification to VerifyPrepareRequest callback

We need to verify the proposed parent's witness in VerifyPrepareRequest, otherwise
the proposed parent may be invalid. Other PrepareRequest fields verification will
be done within the scope of #53.

Also, properly reinitialize sealing proposal on backup nodes after PrepareRequest
verification. This scheme won't work with non-empty blocks after the first CV, but we
have #54 for that.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: replace P2P service with a single callback

We don't need the whole service to be available from DBFT instance, we use
only broadcast callback.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: initialize sealing proposal only if no verification errors occurs

Initialize sealing proposal in the end of new task verification.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: ensure unlocked signer is an authorized validator node

If signer isn't included into validators list for the subsequent block, then
no need to start dBFT process for this block.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: route all incoming txs to dBFT without filtering

dBFT library itself has perfect transaction filtering, thus, no need to spend
resources and create one more layer of filters.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: unify names of setters for dependant services

No functional changes.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: properly manage headers/txs subscriptions

Subscribe, unsubscribe and handle error channel properly.
A part of #29.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: temporary disable difficulty verification

Difficulty calculations must take into account in-turn and out-of-turn
primary, and currently there's no mechanism to sync this dBFT information
with miner proposals. It will be done in a separate task, ref. #58

Also, ensure that stand by committee set via dBFT config is sorted by address.
Pick Primary and NextConsensus from validators sorted by address bytes (not
by votes bytes).

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: remove transaction requester related TODO

Transaction request mechanism will be implemented in a separate PR, ref. #59.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
@roman-khimov, need your review. There are 8 TODOs left (light ones, but still need a bit of thinking and care), will fix them tomorrow and create all appropriate issues.

AnnaShaleva committed:
dbft: do not attach sealing receipts to PrepareRequest

It's just not needed, because receipts root is in the block header, and
all the transactions are available for receipts root checking. The check
itself will be implemented in a separate PR.

The rest of unnecessary fields will be excluded in #56.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: do not optimize parent extra field of PrepareRequest

Later we'll replace first 32 bytes of Extra (extra vanity) with 1 byte
of Extra version, thus it's better to keep the the whole parent Extra (hashable
and unhashable) for proper parent extra bytes handling in VerifyPrepareRequest
callback. Ref. #50.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: temporary go with miner's sealing proposal timestamp

Do not customize timestamp of PrepareRequest with dBFT's value, temporary
go further with timestamp that miner gives to dBFT. The timestamp will be
properly fetched from dBFT in a separate PR, ref. #55.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: skip all encoding constraints for consensus messages

Do not check any of block/transactions/consensus messages restrictions
are met during consensus payload deserialisation. All checks will be
implemented as a part of separate PR either on decoding level (#57) or on
consensus message verification level (#53).

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: directly insert newly added block into chain

The way of insertion is OK, the only thing that may be optimized may be
done after dBFT context initialization refactoring (ref. #53 and #60). Once
we have verified set of receipts in VerifyPrepareRequest callback, we may
store this set of receipts and insert block in chain with the known and
calculated receipts.

This optimisation will be discussed in a separate issue, ref. #60.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: temporary go with uint32 block heigh

Proper dBFT interface extension may be done later if needed, ref. #52.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: add a comment about validators addresses sort order

Validators addresses are sorted and retrieving author using PrimaryIndex
and validators addresses got from block's Extra works correctly now.
This issue doesn not require additional actions, however, we have a lot
of unnecessary fields in dBFT snapshot, this will be improved in #51.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: remove comment about separate memetype request

It will be done in a separate issue. Ref. #61.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
No TODOs left, I'm starting issues creation (I'll add proper links to the commit headers, thus, commit headers are not yet in their final state and will be updated in a few hours).

chenquanyu commented on core/gen_genesis.go:
```diff
@@ -6,6 +6,7 @@ import (
 	"encoding/json"
 	"errors"
 	"math/big"
+	"slices"
```
It seems "slices" package need go version 1.21, we config 1.20 in go.mod file.

roman-khimov replied:
`golang.org/x/exp/slices` can help.

AnnaShaleva replied:
Nice catch, but actually this file is auto-generated, we'd better avoid its modification. Will move to other place.

AnnaShaleva replied:
Moved to another place: one is Genesis block creation, another one is DBFT service constructor. Note, that we then still have unsorted list stored in the DB, but it's OK and doesn't affect node work.

chenquanyu commented on consensus/dbft/dbft.go:
```diff
@@ -82,7 +86,11 @@ var (
-const extraSeal = crypto.SignatureLength // Fixed number of extra-data suffix bytes reserved for a single signer seal
+const (
+	extraSeal = crypto.SignatureLength // Fixed number of extra-data suffix bytes reserved for a single signer seal
+	// txSubCap is the capasity of channel that receives transaction notifications from mempool.
```
Misspell capacity

AnnaShaleva replied:
Fixed.

roman-khimov commented on consensus/dbft/chainreader.go:
```diff
@@ -0,0 +1,22 @@
+	GetBlockByNumber(uint64) *types.Block
+}
+
+type ChainHeaderWriter interface {
```
Exported with no comment.

AnnaShaleva replied:
Fixed.

roman-khimov commented on consensus/dbft/payload.go:
```diff
@@ -0,0 +1,243 @@
+		ValidatorIndex byte
+		ViewNumber     byte
+
+		// relay on default Geth serialization
```
`rely`? But you seem to be using Neo serialization here.

AnnaShaleva replied:
Comment is removed, and we have #62 for it to fix.

roman-khimov commented on consensus/dbft/payload.go:
```diff
@@ -0,0 +1,243 @@
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/nspcc-dev/dbft/payload"
+	"github.com/nspcc-dev/neo-go/pkg/io"
+
```
:eyes:

AnnaShaleva replied:
Fixed.

AnnaShaleva replied:
I thought I finally managed to properly set goland, but seems not.

roman-khimov commented on core/gen_genesis.go:
```diff
@@ -83,6 +84,9 @@ func (g *Genesis) UnmarshalJSON(input []byte) error {
 	if dec.Config != nil {
 		g.Config = dec.Config
+		if g.Config.DBFT != nil && len(g.Config.DBFT.StandByCommittee) != 0 {
+			slices.SortFunc(g.Config.DBFT.StandByCommittee, common.Address.Cmp)
```
I'm not sure this is the best place to sort these keys. Config is a config, we better adjust the order where we care about it.

AnnaShaleva replied:
Fixed. Moved to another place: one is Genesis block creation, another one is DBFT service constructor. Note, that we then still have unsorted list stored in the DB, but it's OK and doesn't affect node work.

roman-khimov commented on consensus/dbft/payload.go:
```diff
@@ -0,0 +1,243 @@
+	recoveryRequestType messageType = 0x40
+	recoveryMessageType messageType = 0x41
+
+	payloadGasLimit = 2000000 // 0.02 GAS
```
What's the meaning of it here?

AnnaShaleva replied:
Removed.

AnnaShaleva committed:
dbft: remove unnecessary comment

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -233,11 +255,19 @@ type DBFT struct {
 func New(config *params.DBFTConfig, db ethdb.Database) *DBFT {
 	// Set any missing consensus parameters to their defaults
 	conf := *config
+	// Sort validators once to reuce the sorted list in getNextConsensus.
```
reuse

AnnaShaleva replied:
Fixed.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -249,12 +279,22 @@ func New(config *params.DBFTConfig, db ethdb.Database) *DBFT {
-		finished:   make(chan struct{}),
+		blockQueue: newBlockQueue(),
+
+		messages:        make(chan Payload, 100),
```
Magic :100:?

AnnaShaleva replied:
No, it's ported from NeoGo and I believe there's a reason for it to be 100. I've created a dedicated constant for it, do we need to create an issue for tuning?

roman-khimov replied:
Ah, NeoGo. Then it's perfect, obviously!

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -324,27 +361,40 @@ func New(config *params.DBFTConfig, db ethdb.Database) *DBFT {
-			// Do not call postBlock here to avoid race between the next pending sealing request that
-			// wants to update the same values as postBlock. Call postBlock in Seal instead.
+			// Avoid copying and may safely change the block itself, as this part
+			// of code is guaranteed to be called once thanks to condition above.
```
I'd add "lastIndex is updated in postBlock", otherwise it's a bit unobvious.

AnnaShaleva replied:
Fixed.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -392,10 +457,109 @@ func New(config *params.DBFTConfig, db ethdb.Database) *DBFT {
+					return fmt.Errorf("invalid parent: expected %s, got %s", c.dbft.PrevHash, req.SealingProposal.ParentHash)
+				}
+				if req.ParentSealHash != c.lastBlockSealHash {
+					return fmt.Errorf("invalid parent: both hash and seal hash mismatch")
```
parent seal hash doesn't match the last block?

AnnaShaleva replied:
Fixed.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -892,104 +1063,253 @@ func (c *DBFT) Seal(chain consensus.ChainHeaderReader, b *types.Block, results c
+		case b := <-c.chainHeadEvents:
+			c.handleChainBlock(b.Block)
+
+			// System has stopped.
```
:eyes:

AnnaShaleva replied:
It  was a comment to receive from subscriptions error channels, but probably it looks a bit unobvious. Adjusted.

AnnaShaleva committed:
dbft: move dBFT messages channel capacity to a dedicated constant

Set it to the same value as NeoGo, it can be changed if needed.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: remove unnecessary comment

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>
